internal/mesh/internal/controllers/sidecarproxy/builder/expose_paths.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package builder
+
+import (
+	"fmt"
+	"regexp"
+
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v2beta1"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v2beta1"
+	"github.com/hashicorp/consul/proto-public/pbmesh/v2beta1/pbproxystate"
+)
+
+func (b *Builder) buildExposePaths(workload *pbcatalog.Workload) {
+	if b.proxyCfg.GetDynamicConfig() != nil && b.proxyCfg.GetDynamicConfig().GetExposeConfig() != nil {
+		for _, exposePath := range b.proxyCfg.GetDynamicConfig().GetExposeConfig().GetExposePaths() {
+			clusterName := exposePathClusterName(exposePath)
+
+			b.addExposePathsListener(workload, exposePath).
+				addExposePathsRouter(exposePath).
+				buildListener()
+
+			b.addExposePathsRoute(exposePath, clusterName).
+				addLocalAppCluster(clusterName).
+				addLocalAppStaticEndpoints(clusterName, exposePath.LocalPathPort)
+		}
+	}
+}
+
+func (b *Builder) addExposePathsListener(workload *pbcatalog.Workload, exposePath *pbmesh.ExposePath) *ListenerBuilder {
+	listenerName := fmt.Sprintf("exposed_path_%s", exposePathName(exposePath))
+
+	listener := &pbproxystate.Listener{
+		Name:      listenerName,
+		Direction: pbproxystate.Direction_DIRECTION_INBOUND,
+	}
+
+	meshAddress := workload.GetFirstNonExternalMeshAddress()
+	if meshAddress == nil {
+		return b.NewListenerBuilder(nil)
+	}
+
+	listener.BindAddress = &pbproxystate.Listener_HostPort{
+		HostPort: &pbproxystate.HostPortAddress{
+			Host: meshAddress.Host,
+			Port: exposePath.LocalPathPort,
+		},
+	}
+
+	return b.NewListenerBuilder(listener)
+}
+
+func (b *ListenerBuilder) addExposePathsRouter(exposePath *pbmesh.ExposePath) *ListenerBuilder {
+	if b.listener == nil {
+		return b
+	}
+	destinationName := exposePathDestinationName(exposePath)
+
+	var l7Protocol pbproxystate.L7Protocol
+
+	switch exposePath.Protocol {
+	case pbmesh.ExposePathProtocol_EXPOSE_PATH_PROTOCOL_HTTP:
+		l7Protocol = pbproxystate.L7Protocol_L7_PROTOCOL_HTTP
+	case pbmesh.ExposePathProtocol_EXPOSE_PATH_PROTOCOL_HTTP2:
+		l7Protocol = pbproxystate.L7Protocol_L7_PROTOCOL_HTTP2
+	default:
+		panic("unsupported expose paths protocol")
```

Comment:
It looks like there aren't validations for `ProxyConfiguration` resources yet, but we definitely should add them soon to avoid a panic here.

Reply:
Yep, we will add them in an upcoming PR.